internal/service/alarms/internal/notifier_provider/notification.go

```diff
@@ -0,0 +1,92 @@
+package notifier_provider
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/google/uuid"
+	"github.com/openshift-kni/oran-o2ims/internal/service/alarms/internal/db/models"
+	a "github.com/openshift-kni/oran-o2ims/internal/service/alarms/internal/db/repo"
+	"github.com/openshift-kni/oran-o2ims/internal/service/common/notifier"
+)
+
+// Compile time check for interface compliance
+var _ notifier.NotificationProvider = (*NotificationStorageProvider)(nil)
+
+// NotificationStorageProvider implements the NotificationProvider interface as a means to abstract the concrete
+// notification type out of the Notifier
+type NotificationStorageProvider struct {
+	repository    *a.AlarmsRepository
+	globalCloudID uuid.UUID
+}
+
+// NewNotificationStorageProvider creates a new NotificationProvider
+func NewNotificationStorageProvider(repository *a.AlarmsRepository) notifier.NotificationProvider {
+	return &NotificationStorageProvider{
+		repository: repository,
+	}
+}
+
+func (n *NotificationStorageProvider) GetNotifications(ctx context.Context) ([]notifier.Notification, error) {
+	var notifications []notifier.Notification
+	// Get all subscriptions
+	subscriptions, err := n.repository.GetAlarmSubscriptions(ctx)
+	if err != nil {
+		return notifications, fmt.Errorf("failed to get all subscriptions for alarms: %w", err)
+	}
+	if len(subscriptions) == 0 {
+		slog.Info("No subscriptions to notify")
+		return notifications, nil
+	} else {
+		slog.Info("Found subscriptions to notify", "subscription count", len(subscriptions))
+		// Find the min event cursor, this will decide the total events that needs to be there ignoring the filter
+		alarms, err := n.repository.GetAlarmsForSubscription(ctx, *getMinSubscription(subscriptions))
+		if err != nil {
+			return notifications, fmt.Errorf("failed to get alarms for subscription: %w", err)
+		}
+
+		// Convert alert to notification and using notificationID as subscriberID
+		for _, sub := range subscriptions {
```

Comment:
Could you maybe add a `TODO` here about what we discussed yesterday regarding needing to refactor this a bit so that we aren't directly cloning the notifications for each subscription and instead have this handled a bit more generically directly in the notifier/worker via an interface callback to insert the consumer subscription id directly into the payload before sending.

As I said, I'm quite happy/willing to do that refactoring since I'm the one that missed this issue with the consumer subscription id!!

Reply:
Yeah i got one todo under Match! Will add a note there too!

Reply:
updated